File: MovieVerse-Mobile/www/utilities/ios-utils/SignInService.swift
import Foundation

class SignInService {
    static let shared = SignInService()

    private init() {}

    func signIn(username: String, password: String, completion: @escaping (Result<User, SignInError>) -> Void) {
        let url = URL(string: "https://MovieVerse.com/signin")! // Replace with your real URL - Mine is not published here
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["username": username, "password": password]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data, error == nil else {
                completion(.failure(.networkError))
                return
            }

            do {
                let user = try JSONDecoder().decode(User.self, from: data)
                completion(.success(user))
            }
            catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        SignInService.shared.signIn(username: "username", password: "password") { result in
            switch result {
            case .success(let user):
                print("Signed in as \(user.name)")
            case .failure(let error):
                print("Error signing in: \(error)")
            }
        }
    }

    private func signIn(username: String, password: String, completion: @escaping (Result<User, SignInError>) -> Void) {
        let url = URL(string: "https://MovieVerse.com/signin")! // Replace with your real URL - Mine is not published here
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["username": username, "password": password]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data, error == nil else {
                completion(.failure(.networkError))
                return
            }

            do {
                let user = try JSONDecoder().decode(User.self, from: data)
                completion(.success(user))
            }
            catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }
}

struct User: Codable {
    let id: String
    let name: String
    let profilePictureURL: URL
    let token: String
    let password: String
}

enum SignInError: Error {
    case networkError
    case decodingError
    case invalidCredentials
    case other
}

File: MovieVerse-Mobile/www/app/android/SignInService.kt
import android.util.Log
import okhttp3.*
import org.json.JSONObject
import java.io.IOException

object SignInService {
    private const val URL = "https://MovieVerse.com/signin"

    fun signIn(username: String, password: String, callback: (Result<User>) -> Unit) {
        val client = OkHttpClient()

        val requestBody = FormBody.Builder()
            .add("username", username)
            .add("password", password)
            .build()

        val request = Request.Builder()
            .url(URL)
            .post(requestBody)
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(e))
            }

            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(IOException("Unexpected code $response")))
                    return
                }
                else if (response.body == null) {
                    callback(Result.failure(IOException("Empty response body")))
                    return
                }

                try {
                    val jsonObject = JSONObject(response.body?.string() ?: "")
                    val user = User(
                        id = jsonObject.getString("id"),
                        name = jsonObject.getString("name"),
                        email = jsonObject.getString("email"),
                        password = jsonObject.getString("password"),
                        token = jsonObject.getString("token")
                    )
                    callback(Result.success(user))
                }
                catch (e: Exception) {
                    callback(Result.failure(e))
                }
            }
        })
    }
}

data class User(
    val id: String,
    val name: String,
    val email: String,
    val password: String,
    val token: String
)
